src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -822,17 +822,87 @@ internal override TypeSymbol SetUnknownNullabilityForReferenceTypes()
                 }
             }
 
-            if (!haveChanges)
+            TypeSymbol result = this;
+            if (haveChanges)
             {
-                allTypeArguments.Free();
-                return this;
+                var definition = this.OriginalDefinition;
+                TypeMap substitution = new TypeMap(definition.GetAllTypeParameters(), allTypeArguments.ToImmutable());
+                result = substitution.SubstituteNamedType(definition);
             }
-            else
+
+            allTypeArguments.Free();
+            return result;
+        }
+
+        internal override (bool, TypeSymbol) TryMergeNullability(TypeSymbol other, VarianceKind variance)
+        {
+            Debug.Assert(this.Equals(other, TypeCompareKind.IgnoreDynamicAndTupleNames));
+
+            if (!IsGenericType)
             {
-                TypeMap substitution = new TypeMap(this.OriginalDefinition.GetAllTypeParameters(),
-                                                   allTypeArguments.ToImmutableAndFree());
+                return (true, this);
+            }
+
+            var definition = this.OriginalDefinition;
+            var allTypeParameters = definition.GetAllTypeParameters();
+            var allTypeArguments = ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance();
```

Comment:
Can we not specify starting sizes?

Reply:
The size will be 1 or 2 typically so that might not have any effect.

---
In reply to: [223862549](https://github.com/dotnet/roslyn/pull/29727#discussion_r223862549) [](ancestors = 223862549)